### PR TITLE
[WIP] Feature/uniswap v3

### PR DESCRIPTION
**Pull request by kavsky02**

### CODE REVIEW

File: contracts/facets/avalanche/TraderJoeV2Facet.sol [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    function addLiquidityTraderJoeV2(ILBRouter.LiquidityParameters memory liquidityParameters) external nonReentrant onlyOwner noBorrowInTheSameBlock recalculateAssetsExposure remainsSolvent {
+        ILBRouter traderJoeV2Router = ILBRouter(JOE_V2_ROUTER_ADDRESS);
+        TraderJoeV2Bin[] memory ownedBins = getOwnedBins();
+
+        liquidityParameters.to = address(this);
+        liquidityParameters.refundTo = address(this);
+
+        address(liquidityParameters.tokenX).safeApprove(address(traderJoeV2Router), liquidityParameters.amountX);
```
Comment (kavsky02) [inline contracts/facets/avalanche/TraderJoeV2Facet.sol:39]: safeApprove 0 first?

File: contracts/facets/avalanche/TraderJoeV2Facet.sol [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        assembly{
+            result.slot := sload(slot)
+        }
+    }
+
+    function getOwnedTraderJoeV2Bins() public view returns (TraderJoeV2Bin[] memory result){
+        return getOwnedBins();
+    }
+
+    function addLiquidityTraderJoeV2(ILBRouter.LiquidityParameters memory liquidityParameters) external nonReentrant onlyOwner noBorrowInTheSameBlock recalculateAssetsExposure remainsSolvent {
```
Comment (kavsky02) [inline contracts/facets/avalanche/TraderJoeV2Facet.sol:32]: Should we conditionally (balance == 0) remove owned assets (tokenX, tokenY) during this operation?

File: contracts/facets/avalanche/TraderJoeV2Facet.sol [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+                }
+            }
+
+            if (!userHadBin) {
+                getOwnedBins().push(TraderJoeV2Bin(pairInfo.LBPair, uint24(depositIds[i])));
+            }
+        }
+    }
+
+    function removeLiquidityTraderJoeV2(RemoveLiquidityParameters memory parameters) external nonReentrant onlyOwnerOrInsolvent noBorrowInTheSameBlock recalculateAssetsExposure remainsSolvent {
```
Comment (kavsky02) [inline contracts/facets/avalanche/TraderJoeV2Facet.sol:67]: Should we conditionally (balance == 0) add owned assets (tokenX, tokenY) during this operation?

File: contracts/facets/SolvencyFacetProd.sol [hunk 4]
```diff
@@ -429,13 +443,167 @@ contract SolvencyFacetProd is AvalancheDataServiceConsumerBase, DiamondHelper {
             if (success) {
                 uint256 balance = abi.decode(result, (uint256));
                 IERC20Metadata token = IERC20Metadata(DeploymentConstants.getTokenManager().getAssetAddress(stakedPositionsPrices[i].asset, true));
-                usdValue += stakedPositionsPrices[i].price * 10 ** 10 * balance / (10 ** token.decimals());
+                usdValue += stakedPositionsPrices[i].price * 10 ** 8 * balance / (10 ** token.decimals());
```
Comment (kavsky02) [inline contracts/facets/SolvencyFacetProd.sol:446]: Why this change? Probably we want to revert it back :)

File: contracts/facets/SolvencyFacetProd.sol [hunk 5]
```diff
@@ -429,13 +443,167 @@ contract SolvencyFacetProd is AvalancheDataServiceConsumerBase, DiamondHelper {
...
 
+    /**
+     **/
+    function getTotalTraderJoeV2() public view virtual returns (uint256) {
+        return getTotalTraderJoeV2WithPrices();
+    }
+
+    /**
+    **/
+    function _getTotalTraderJoeV2WithPricesBase(bool weighted) internal view returns (uint256) {
```
Comment (kavsky02) [inline contracts/facets/SolvencyFacetProd.sol:461]: Seems like there is no prices object being passed here even though the name suggests otherwise

File: contracts/facets/SolvencyFacetProd.sol [hunk 6]
```diff
@@ -429,13 +443,167 @@ contract SolvencyFacetProd is AvalancheDataServiceConsumerBase, DiamondHelper {
...
+                    symbols[0] = DeploymentConstants.getTokenManager().tokenAddressToSymbol(address(binInfo.pair.getTokenX()));
+                    symbols[1] = DeploymentConstants.getTokenManager().tokenAddressToSymbol(address(binInfo.pair.getTokenY()));
+
+                    prices = getOracleNumericValuesFromTxMsg(symbols);
+                }
+
+                {
+                    (uint128 binReserveX, uint128 binReserveY) = binInfo.pair.getBin(binInfo.id);
+
+                    price = PriceHelper.convert128x128PriceToDecimal(binInfo.pair.getPriceFromId(binInfo.id)); // how is it denominated (what precision)?
```
Comment (kavsky02) [inline contracts/facets/SolvencyFacetProd.sol:499]: would be good to ask TJ directly in the chat to avoid taking the chances to guess

File: contracts/facets/SolvencyFacetProd.sol [hunk 7]
```diff
@@ -429,13 +443,167 @@ contract SolvencyFacetProd is AvalancheDataServiceConsumerBase, DiamondHelper {
...
+                }
+
+
+                {
+                    uint256 debtCoverageX = weighted ? DeploymentConstants.getTokenManager().debtCoverage(address(binInfo.pair.getTokenX())) : 1e18;
+                    uint256 debtCoverageY = weighted ? DeploymentConstants.getTokenManager().debtCoverage(address(binInfo.pair.getTokenY())) : 1e18;
+
+                    total = total +
+                    Math.min(
+                        debtCoverageX * liquidity / 10 ** IERC20Metadata(address(binInfo.pair.getTokenY())).decimals() * prices[0] / (price * 10 ** 2),
```
Comment (kavsky02) [inline contracts/facets/SolvencyFacetProd.sol:516]: I guess that this is taking "the worst case scenario", right? Why are we dividing by price * 10 ** 2 in first case but then / 10 ** 8 in the second case?

File: contracts/facets/avalanche/UniswapV3Facet.sol [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+
+    function mintLiquidityUniswapV3(INonfungiblePositionManager.MintParams memory params) external nonReentrant onlyOwner noBorrowInTheSameBlock recalculateAssetsExposure remainsSolvent {
+        address poolAddress = PoolAddress.computeAddress(UNISWAP_V3_FACTORY_ADDRESS, PoolAddress.getPoolKey(params.token0, params.token1, params.fee));
+
+        if (!isPoolWhitelisted(poolAddress)) revert UniswapV3PoolNotWhitelisted();
+
+        params.recipient = address(this);
+
+        //TODO: check for max and min ticks
+        address(params.token0).safeApprove(address(NONFUNGIBLE_POSITION_MANAGER_ADDRESS), params.amount0Desired);
```
Comment (kavsky02) [inline contracts/facets/avalanche/UniswapV3Facet.sol:56]: safe approve for 0 first?

File: contracts/facets/avalanche/UniswapV3Facet.sol [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+        INonfungiblePositionManager.CollectParams memory collectParams = INonfungiblePositionManager.CollectParams(params.tokenId, address(this), uint128(amount0), uint128(amount1));
+
+        INonfungiblePositionManager(NONFUNGIBLE_POSITION_MANAGER_ADDRESS).collect(collectParams);
+
+        ITokenManager tokenManager = DeploymentConstants.getTokenManager();
+
+        emit DecreaseLiquidityUniswapV3(msg.sender, poolAddress, params.tokenId, tokenManager.tokenAddressToSymbol(token0), tokenManager.tokenAddressToSymbol(token1), amount0, amount1, block.timestamp);
+    }
+
+    function burnLiquidityUniswapV3(uint256 tokenId) external nonReentrant onlyOwner noBorrowInTheSameBlock recalculateAssetsExposure onlyOwnerOrInsolvent {
```
Comment (kavsky02) [inline contracts/facets/avalanche/UniswapV3Facet.sol:129]: add/remove owned assets?

Comment (mbare0) [reply contracts/facets/avalanche/UniswapV3Facet.sol:129]: in burn it's not needed (it must be empty to burn)

File: contracts/facets/SolvencyFacetProd.sol (hunk 7 above)
Comment (mbare0) [reply contracts/facets/SolvencyFacetProd.sol:504]: Nice catch! I wrongly assumed, that prices are denominated in 10**6, they are denominated in the decimals of the Y token instead. 
So:
-liquidity - denominated in Y
-prices - denominated 10**8 (RedStone)
-price - denominated in Y

That is why we can simplify this part to:
```
                    Math.min(
                        debtCoverageX * liquidity * prices[0] / (price * 10 ** 8),
                        debtCoverageY * liquidity / 10 ** IERC20Metadata(address(binInfo.pair.getTokenY())).decimals() * prices[1] / 10 ** 8
                    )
                    ```

Comment (mbare0) [reply contracts/facets/SolvencyFacetProd.sol:504]: This is the worst case scenario from two extremes - active bin is on the left or on the right from that bin (so this bin is just not the active bin). Need to confirm with TraderJoe that it is a good assumption.

File: contracts/facets/SolvencyFacetProd.sol [hunk 10]
```diff
@@ -429,13 +443,167 @@ contract SolvencyFacetProd is AvalancheDataServiceConsumerBase, DiamondHelper {
...
+
+                {
+                    (uint128 binReserveX, uint128 binReserveY) = binInfo.pair.getBin(binInfo.id);
+
+                    price = PriceHelper.convert128x128PriceToDecimal(binInfo.pair.getPriceFromId(binInfo.id)); // how is it denominated (what precision)?
+
+                    //TODO: check what are limitations of this
+
+                    liquidity = price * binReserveX
+                        / 10 ** (6 + IERC20Metadata(address(binInfo.pair.getTokenX())).decimals())
```
[the remark this answers is not in the document]
Comment (mbare0) [reply contracts/facets/SolvencyFacetProd.sol:492]: I assumed wrongly that price is denominated in 10^6, it is denominated in token Y decimals instead. So the resulting equation is actually simpler:
```
                    liquidity = price * binReserveX
                        / 10 ** IERC20Metadata(address(binInfo.pair.getTokenX())).decimals()
                        + binReserveY;

File: contracts/facets/SolvencyFacetProd.sol (hunk 6 above)
Comment (jakub-wojciechowski) [reply contracts/facets/SolvencyFacetProd.sol:485]: Let's do it, based on our UniV3 experience this seems like a ratio between assets (i.e. how much of asset1 you can get for asset0)
This could be manipulated (e.g.deploying more asset0 tokens to the pool), therefore usually people use TWAP to eliminate the risk of flash-loan attacks.

File: contracts/facets/SolvencyFacetProd.sol [hunk 11]
```diff
@@ -436,6 +448,137 @@ contract SolvencyFacetProd is AvalancheDataServiceConsumerBase, DiamondHelper {
...
+
+                    prices = getOracleNumericValuesFromTxMsg(symbols);
+                }
+
+                {
+                    (uint128 binReserveX, uint128 binReserveY) = binInfo.pair.getBin(binInfo.id);
+
+                    price = PriceHelper.convert128x128PriceToDecimal(binInfo.pair.getPriceFromId(binInfo.id)); // how is it denominated (what precision)?
+
+                    liquidity = price * binReserveX
```
Comment (jakub-wojciechowski) [inline contracts/facets/SolvencyFacetProd.sol:487]: In which unit is the liquidity?

File: contracts/facets/SolvencyFacetProd.sol (hunk 6 above)
Comment (jakub-wojciechowski) [reply contracts/facets/SolvencyFacetProd.sol:485]: Is the price constant per bin or could it change?

File: contracts/facets/avalanche/UniswapV3Facet.sol [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+        params.recipient = address(this);
+
+        //TODO: check for max and min ticks
+        address(params.token0).safeApprove(address(NONFUNGIBLE_POSITION_MANAGER_ADDRESS), 0);
+        address(params.token1).safeApprove(address(NONFUNGIBLE_POSITION_MANAGER_ADDRESS), 0);
+
+        address(params.token0).safeApprove(address(NONFUNGIBLE_POSITION_MANAGER_ADDRESS), params.amount0Desired);
+        address(params.token1).safeApprove(address(NONFUNGIBLE_POSITION_MANAGER_ADDRESS), params.amount1Desired);
+
+        (uint256 tokenId,
```
Comment (jakub-wojciechowski) [inline contracts/facets/avalanche/UniswapV3Facet.sol:67]: We should probably do a sanity check here if decimal of sqrtPrice doesn't deviate from token0/token1 ratio based on redstone Oracles.

File: contracts/facets/SolvencyFacetProd.sol [hunk 13]
```diff
@@ -436,6 +448,137 @@ contract SolvencyFacetProd is AvalancheDataServiceConsumerBase, DiamondHelper {
...
+                        + binReserveY;
+                }
+
+
+                {
+                    uint256 debtCoverageX = weighted ? DeploymentConstants.getTokenManager().debtCoverage(address(binInfo.pair.getTokenX())) : 1e18;
+                    uint256 debtCoverageY = weighted ? DeploymentConstants.getTokenManager().debtCoverage(address(binInfo.pair.getTokenY())) : 1e18;
+
+                    total = total +
+                    Math.min(
```
Comment (jakub-wojciechowski) [inline contracts/facets/SolvencyFacetProd.sol:498]: We should check, if we don't assume to negative scenatio for valuation (e.g. just after user opens a position the valuation is much lower than added funds)

File: contracts/facets/SolvencyFacetProd.sol [hunk 14]
```diff
@@ -436,6 +448,137 @@ contract SolvencyFacetProd is AvalancheDataServiceConsumerBase, DiamondHelper {
...
+
+                    total = total +
+
+                    //TODO: there is an assumption here that the position value is the lowest at edges (x = 0 or y = 0). Need to confirm that!
+
+                    //TODO: tickerUpper = p_b, tickerLower = p_a, first check if that's correct, secondly check what's the denomination and accuracy of these numbers
+                    //TODO: check for possible under/overflowsL557
+
+
+                    Math.min(
```
Comment (jakub-wojciechowski) [inline contracts/facets/SolvencyFacetProd.sol:560]: We should check, if we don't assume to negative scenatio for valuation (e.g. just after user opens a position the valuation is much lower than added funds)

File: contracts/facets/avalanche/TraderJoeV2Facet.sol [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+
+    function addLiquidityTraderJoeV2(ILBRouter.LiquidityParameters memory liquidityParameters) external nonReentrant onlyOwner noBorrowInTheSameBlock recalculateAssetsExposure remainsSolvent {
+        ILBRouter traderJoeV2Router = ILBRouter(JOE_V2_ROUTER_ADDRESS);
+        TraderJoeV2Bin[] memory ownedBins = getOwnedTraderJoeV2Bins();
+        ILBFactory lbFactory = traderJoeV2Router.getFactory();
+        ILBFactory.LBPairInformation memory pairInfo = lbFactory.getLBPairInformation(liquidityParameters.tokenX, liquidityParameters.tokenY, liquidityParameters.binStep);
+
+        if (!isPairWhitelisted(address(pairInfo.LBPair))) revert TraderJoeV2PoolNotWhitelisted();
```
Comment (jakub-wojciechowski) [inline contracts/facets/avalanche/TraderJoeV2Facet.sol:50]: We must prohibit adding liquidity to bins that aren't active based on MARKET (RedStone) and not TraderJoe price.
Let's just make a sanity check if TradeJoe price doesn't equals market price with the precision of a single bin.

**Comment by kavsky02** [comment]
No longer needed
